predicates/traffic/segment.go

```diff
@@ -0,0 +1,77 @@
+package traffic
+
+import (
+	"math/rand"
+	"net/http"
+
+	"github.com/zalando/skipper/predicates"
+	"github.com/zalando/skipper/routing"
+)
+
+type (
+	segmentSpec      struct{}
+	segmentPredicate struct{ min, max float64 }
+)
+
+type contextKey struct{}
+
+var randomValue contextKey
+
+// NewSegment creates a new traffic segment predicate specification
+func NewSegment() routing.WeightedPredicateSpec {
+	return &segmentSpec{}
+}
+
+func (*segmentSpec) Name() string {
+	return predicates.TrafficSegmentName
+}
+
+// Create new predicate instance with two number arguments _min_ and _max_
+// from an interval [0, 1] (from zero included to one included) and _min_ <= _max_.
+//
+// Let _r_ be one-per-request uniform random number value from [0, 1).
+// This predicate matches if _r_ belongs to an interval from [_min_, _max_).
+// Upper interval boundary _max_ is excluded to simplify definition of
+// adjacent intervals - the upper boundary of the first interval
+// then equals lower boundary of the next and so on, e.g. [0, 0.25) and [0.25, 1).
+//
+// This predicate has weight of -1 and therefore does not affect route weight.
+//
+// Example of routes splitting traffic in 50%+30%+20% proportion:
+//
+//	r50: Path("/test") && TrafficSegment(0.0, 0.5) -> <shunt>;
+//	r30: Path("/test") && TrafficSegment(0.5, 0.8) -> <shunt>;
+//	r20: Path("/test") && TrafficSegment(0.8, 1.0) -> <shunt>;
+func (*segmentSpec) Create(args []any) (routing.Predicate, error) {
+	if len(args) != 2 {
+		return nil, predicates.ErrInvalidPredicateParameters
+	}
+
+	p, ok := &segmentPredicate{}, false
+
+	if p.min, ok = args[0].(float64); !ok || p.min < 0 || p.min > 1 {
```

Comment:
How can we stop user for providing invalid input (segments that do not sum up to 1)?
For example like this:
```
r55: Path("/test") && TrafficSegment(0.0, 0.55) -> <shunt>;
r40: Path("/test") && TrafficSegment(0.55, 0.95) -> <shunt>;
```

Reply:
We do not. This predicate is a ["plumbing"](https://git-scm.com/book/en/v2/Git-Internals-Plumbing-and-Porcelain) and meant to be a building block of higher level algorithm that must ensure there no gaps or overlaps between segments.